File: #1095.find-in-mountain-array.cpp
/**
 * // This is the MountainArray's API interface.
 * // You should not implement it, or speculate about its implementation
 * class MountainArray {
 *   public:
 *     int get(int index);
 *     int length();
 * };
 */

class Solution {
public:
    int findInMountainArray(int target, MountainArray &mountainArr) {

        int length = mountainArr.length();
        int left = 0;
        int right = length - 1;
        int peak = -1;
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (mountainArr.get(mid) < mountainArr.get(mid + 1)) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        peak = left;
        if (mountainArr.get(peak) == target) {
            return peak;
        }

        int search1 = search(target, mountainArr, 0, peak - 1, true);
        int search2 = search(target, mountainArr, peak + 1, length - 1, false);
        return search1 == -1 ? search2 : search1;
    }

private:
    int search(int target, MountainArray &mountainArr,
               int left, int right, bool isAsc) {

        while (left < right) {
            int mid = left + (right - left) / 2;
            int index = mountainArr.get(mid);
            if (index == target) {
                return mid;
            }
            
            if (index < target) {
                if (isAsc) {
                    left = mid + 1;
                } else {
                    right = mid;
                }
            } else {
                if (isAsc) {
                    right = mid;
                } else {
                    left = mid + 1;
                }
            }
        }
        return mountainArr.get(right) == target ? right : -1;
    }
};

File: #1095.find-in-mountain-array.js
/**
 * // This is the MountainArray's API interface.
 * // You should not implement it, or speculate about its implementation
 * function MountainArray() {
 *     @param {number} index
 *     @return {number}
 *     this.get = function(index) {
 *         ...
 *     };
 *
 *     @return {number}
 *     this.length = function() {
 *         ...
 *     };
 * };
 */

/**
 * @param {number} target
 * @param {MountainArray} mountainArr
 * @return {number}
 */
var findInMountainArray = function(target, mountainArr) {
    const cache = {};
    const getCachedValue = (i) => {
        if (undefined === cache[i]) {
            cache[i] = mountainArr.get(i);
        }
        return cache[i];
    }

    const binarySearch = (lo, hi, desc) => {
        while (lo <= hi) {
            let mid = Math.trunc(lo + (hi - lo) / 2);
            const val = getCachedValue(mid);
            if (val == target)  {
                return mid;
            } else if (desc && val > target || !desc && val <= target) {
                lo = mid + 1;
            } else {
                hi = mid - 1;
            }
        }
        return -1;
    }

    let max = 0, n = mountainArr.length();
    cache[n] = Number.MIN_VALUE;
    let right = n;
    while (max < right) {
        let mid = Math.trunc(max + (right - max) / 2);
        if (getCachedValue(mid) > getCachedValue(mid + 1)) {
            right = mid;
        } else {
            max = mid + 1;
        }
    }

    const result = binarySearch(0, max, false);
    return result == -1 ? binarySearch(max, n - 1, true) : result;
};
